homeassistant/components/sensor/teksavvy.py

```diff
@@ -0,0 +1,165 @@
+    'onpeak_remaining': ['Remaining', GIGABYTES, 'mdi:download']
+}
+
+API_HA_MAP = (
+    ('OnPeakDownload', 'onpeak_download'),
+    ('OnPeakUpload', 'onpeak_upload'),
+    ('OffPeakDownload', 'offpeak_download'),
+    ('OffPeakUpload', 'offpeak_upload'))
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_MONITORED_VARIABLES):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_TOTAL_BANDWIDTH): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the sensor platform."""
+    apikey = config.get(CONF_API_KEY)
+    bandwidthcapstr = config.get(CONF_TOTAL_BANDWIDTH)
+
+    try:
+        bandwidthcap = int(bandwidthcapstr)
+        teksavvy_data = TekSavvyData(apikey, bandwidthcap)
+        teksavvy_data.update()
+    except ValueError as error:
+        _LOGGER.error("Failed conversion %s %s", CONF_TOTAL_BANDWIDTH, error)
+    except requests.exceptions.HTTPError as error:
+        _LOGGER.error("Fail to fetch data %s", error)
+        return False
+
+    name = config.get(CONF_NAME)
+
+    sensors = []
+    for variable in config[CONF_MONITORED_VARIABLES]:
+        sensors.append(TekSavvySensor(teksavvy_data, variable, name))
+
+    add_devices(sensors, True)
+
+
+class TekSavvySensor(Entity):
+    """TekSavvy Bandwidth sensor."""
+
+    def __init__(self, teksavvydata, sensor_type, name):
+        """Initialize the sensor."""
+        self.client_name = name
+        self.type = sensor_type
+        self._name = SENSOR_TYPES[sensor_type][0]
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+        self._icon = SENSOR_TYPES[sensor_type][2]
+        self.teksavvydata = teksavvydata
+        self._state = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return '{} {}'.format(self.client_name, self._name)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    def update(self):
+        """Get the latest data from TekSavvy and update the state."""
+        self.teksavvydata.update()
+        if self.type in self.teksavvydata.data:
+            self._state = round(self.teksavvydata.data[self.type], 2)
+
+
+class TekSavvyData(object):
+    """Get data from TekSavvy API."""
+
+    def __init__(self, api_key, bandwidth_cap):
+        """Initialize the data object."""
+        self.api_key = api_key
+        self.bandwidth_cap = bandwidth_cap
+        self.data = dict()
+        self.data["limit"] = self.bandwidth_cap
+
+    def _fetch_data(self):
```

Comment:
Docstrings are required everywhere.

Reply:
Done. Would be nice if we could add that to lint.